test/cast/discriminator_test.exs

```diff
@@ -0,0 +1,215 @@
+defmodule OpenApiSpex.CastDiscriminatorTest do
+  use ExUnit.Case
+  alias OpenApiSpex.{Cast, Schema}
+  alias OpenApiSpex.Cast.{Discriminator, Error}
+
+  # The discriminator we'll be using across the tests. Animal type was
+  # specifically chosen to help the examples be more clear since abstract
+  # objects can be difficult to grok in this context.
+  @discriminator "animal_type"
+
+  defp cast(ctx) do
+    Discriminator.cast(struct(Cast, ctx))
+  end
+
+  defp cast_cast(ctx) do
+    Cast.cast(struct(Cast, ctx))
+  end
+
+  # Maps an arbitrary string value to a schema
+  def build_discriminator_mapping(name, schema) do
+    %{name => schema_ref(schema)}
+  end
+
+  def schema_ref(%{title: title}), do: "#/components/schemas/#{title}"
+
+  def build_schema(title, properties) do
+    %Schema{
+      type: :object,
+      title: title,
+      properties: properties
+    }
+  end
+
+  def build_discriminator_schema(schemas, composite_keyword, property_name, mapppings \\ nil) do
+    %Schema{
+      type: :object,
+      discriminator: %{
+        propertyName: "#{property_name}",
+        mapping: mapppings
+      }
+    }
+    |> Map.put(composite_keyword, schemas)
+  end
+
+  setup do
+    dog_schema =
+      build_schema("Dog", %{
+        breed: %Schema{type: :string},
+        age: %Schema{type: :integer}
+      })
+
+    wolf_schema =
+      build_schema("Wolf", %{
+        breed: %Schema{type: :string, minLength: 5},
+        age: %Schema{type: :integer}
+      })
+
+    cat_schema =
+      build_schema("Cat", %{
+        breed: %Schema{type: :string},
+        lives: %Schema{type: :integer}
+      })
+
+    {:ok, schemas: %{dog: dog_schema, cat: cat_schema, wolf: wolf_schema}}
+  end
+
+  describe "cast/1" do
+    test "basics, anyOf", %{schemas: %{dog: dog, wolf: wolf}} do
+      # "animal_type" is the discriminator and the keys need to be strings.
+      input_value = %{@discriminator => "Dog", "breed" => "Pug", "age" => 1}
+      # Create the discriminator schema. Discriminators require a composite
+      # key be provided (`allOf`, `anyOf`, `oneOf`).
+      discriminator_schema =
+        build_discriminator_schema([dog, wolf], :anyOf, String.to_atom(@discriminator), nil)
+
+      # Note: We're expecting to getting atoms back, not strings
+      expected = {:ok, %{age: 1, breed: "Pug"}}
+
+      assert expected == cast(value: input_value, schema: discriminator_schema)
```

Comment:
Usually, the expected value is on the right side of the `==`. It's probably a good thing to be consistent about that.

Reply:
```elixir
cast_result = cast(value: input_value, schema: discriminator_schema)

# Note: We're expecting to getting atoms back, not strings
assert cast_result == {:ok, %{age: 1, breed: "Pug"}}
```